module/data/fields/collection-field.ts

```diff
@@ -0,0 +1,105 @@
+import { AnyObject } from 'fvtt-types/utils'
+import fields = foundry.data.fields
+import DataModel = foundry.abstract.DataModel
+import { PseudoDocument } from '../../pseudo-document/pseudo-document.js'
+import { TypedPseudoDocument } from '../../pseudo-document/typed-pseudo-document.js'
+import { ModelCollection } from '../model-collection.js'
+
+class LazyTypedSchemaField<
+  const Types extends fields.TypedSchemaField.Types,
+  const Options extends fields.TypedSchemaField.Options<Types> = fields.TypedSchemaField.DefaultOptions,
+  // eslint-disable-next-line @typescript-eslint/no-deprecated
+  const AssignmentType = fields.TypedSchemaField.AssignmentType<Types, Options>,
+  const InitializedType = fields.TypedSchemaField.InitializedType<Types, Options>,
+  const PersistedType = fields.TypedSchemaField.PersistedType<Types, Options>,
+> extends fields.TypedSchemaField<Types, Options, AssignmentType, InitializedType, PersistedType> {
+  protected override _validateSpecial(value: AssignmentType): boolean | void {
+    if (!value || (value as any).type in this.types) return super._validateSpecial(value)
+    return true
+  }
+}
+
+/* ---------------------------------------- */
+
+namespace CollectionField {
+  export type Types<Model extends typeof PseudoDocument> = {
```

Comment:
If this is hardcoded tied to PseudoDocument, you might as well move it to that directory. I would expect everything in a top level data folder to be generic and reusable.